### PR TITLE
Use native encoded bytes when calling native printf

### PR DESCRIPTION
**Pull request by r30shah**
When making a downcall to standard `printf()`, use native encoded bytes as argument instead of UTF-8 bytes.

### CODE REVIEW

**Comment by r30shah** [comment]
@keithc-ca  Can I get your view on this change ? This change to make sure when test is calling to standard printf, it uses native encoding instead of UTF-8 which will not be native encoding on all platforms.

**Comment by keithc-ca** [comment]
This should first be proposed for https://github.com/ibmruntimes/openj9-openjdk-jdk, then https://github.com/ibmruntimes/openj9-openjdk-jdk23 before this version.

**Comment by r30shah** [comment]
I am looking into this and should have mentioned,
I started making update to jdk first,  but that test was updated with new MemorySegment apis that is not available in JDK21 [1][2]. From the looks of it. I am assessing if the way test was modified, would work on z/OS or not (I think it will still have same issue as by default the String with which it will call printf is going to be UTF-8. But I need to confirm this.

Marking this WIP while I confirm this. 

[1]. https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/a4feda773523caf5471d7aa5c29146b350d993d5/test/jdk/java/foreign/StdLibTest.java#L120-L136
[2]. https://github.com/ibmruntimes/openj9-openjdk-jdk23/blob/bad8ef276348989b6ff809cab13f8ae1e281bdf3/test/jdk/java/foreign/StdLibTest.java#L120-L136

**Comment by keithc-ca** [comment]
> new MemorySegment apis

Yes, the APIs have changed, but I believe all the functionality that would be needed is available in each version.

File: test/jdk/java/foreign/StdLibTest.java [hunk 1]
```diff
@@ -300,7 +318,7 @@ int rand() throws Throwable {
 
         int printf(String format, List<PrintfArg> args) throws Throwable {
             try (var arena = Arena.ofConfined()) {
-                MemorySegment formatStr = arena.allocateUtf8String(format);
+                MemorySegment formatStr = arena.allocateArray(ValueLayout.JAVA_BYTE, getNativeEncodingBytesFromString(format));
```
Comment (r30shah) [inline test/jdk/java/foreign/StdLibTest.java:321]: Given that JDK21 does not have MemorySegment API that can allocate segment from String with passed charset, for JDK21, we need to create Segment from array. I have implemented the fix here to use array. 

Changes have been tested on AIX (`/view/Test_grinder/job/Grinder/43870`) and S390 Linux (`/view/Test_grinder/job/Grinder/43869`):

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: test/jdk/java/foreign/StdLibTest.java [hunk 2]
```diff
@@ -58,6 +65,17 @@ public class StdLibTest extends NativeTestHelper {
...
+    static byte[] getNativeEncodingBytesFromString(String str) {
+        byte[] nativeBytes;
+        try {
+            nativeBytes = str.getBytes(System.getProperty("native.encoding"));
+        } catch (UnsupportedEncodingException ex) {
+            // Return bytes with default charset
+            nativeBytes = str.getBytes();
+        }
+        return Arrays.copyOf(nativeBytes, nativeBytes.length + 1);
+    }
```
[the remark this answers is not in the document]
Comment (r30shah) [reply test/jdk/java/foreign/StdLibTest.java]: Thanks a lot, this does look more clean - made changes in https://github.com/ibmruntimes/openj9-openjdk-jdk21/compare/7870dd85edbe92994a022b274914c9b726a95b56..960e06a3679336e039a146ae1ebe4fcfa56da55c

File: test/jdk/java/foreign/StdLibTest.java [hunk 3]
```diff
@@ -31,6 +37,8 @@
 import java.lang.invoke.MethodHandle;
 import java.lang.invoke.MethodHandles;
 import java.lang.invoke.MethodType;
+import java.nio.charset.Charset;
+import java.io.UnsupportedEncodingException;
```
Comment (keithc-ca) [inline test/jdk/java/foreign/StdLibTest.java:41]: Please remove this unused import.

File: test/jdk/java/foreign/StdLibTest.java [hunk 4]
```diff
@@ -58,6 +66,13 @@ public class StdLibTest extends NativeTestHelper {
 
     private StdLibHelper stdLibHelper = new StdLibHelper();
 
+    final static Charset nativeCharset = Charset.forName(System.getProperty("native.encoding"));
+
+    static MemorySegment nativeBytesFromString(Arena arena, String str) {
+        byte[] nativeBytes = str.getBytes(nativeCharset);
+        return arena.allocateArray(ValueLayout.JAVA_BYTE, Arrays.copyOf(nativeBytes, nativeBytes.length + 1));
+    }
```
Comment (keithc-ca) [inline test/jdk/java/foreign/StdLibTest.java:74]: Please move this before line 67 to keep all static features together.

File: test/jdk/java/foreign/StdLibTest.java [hunk 5]
```diff
@@ -300,7 +315,7 @@ int rand() throws Throwable {
 
         int printf(String format, List<PrintfArg> args) throws Throwable {
             try (var arena = Arena.ofConfined()) {
-                MemorySegment formatStr = arena.allocateUtf8String(format);
+                MemorySegment formatStr = arena.allocateArray(ValueLayout.JAVA_BYTE, nativeBytesFromString(arena, format));
```
Comment (keithc-ca) [inline test/jdk/java/foreign/StdLibTest.java:318]: This must change (I don't think it compiles as is):
```java
                MemorySegment formatStr = nativeBytesFromString(arena, format);
```
A similar change is needed on line 399.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: test/jdk/java/foreign/StdLibTest.java (hunk 3 above)
Comment (r30shah) [reply test/jdk/java/foreign/StdLibTest.java]: Removed in https://github.com/ibmruntimes/openj9-openjdk-jdk21/compare/960e06a3679336e039a146ae1ebe4fcfa56da55c..a837b75d57e86bb113c40ee75d4738a6af6e1e8d

File: test/jdk/java/foreign/StdLibTest.java (hunk 4 above)
Comment (r30shah) [reply test/jdk/java/foreign/StdLibTest.java:71]: Moved this in https://github.com/ibmruntimes/openj9-openjdk-jdk21/compare/960e06a3679336e039a146ae1ebe4fcfa56da55c..a837b75d57e86bb113c40ee75d4738a6af6e1e8d

File: test/jdk/java/foreign/StdLibTest.java (hunk 5 above)
Comment (r30shah) [reply test/jdk/java/foreign/StdLibTest.java]: Sorry about this goof-up. When I tested the change to ensure functionality I had it implemented correctly, In pushing changes I used search and replace. I should have been more careful, apologize for this. 

Changed in https://github.com/ibmruntimes/openj9-openjdk-jdk21/compare/960e06a3679336e039a146ae1ebe4fcfa56da55c..a837b75d57e86bb113c40ee75d4738a6af6e1e8d

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**